python/fbprophet/forecaster.py

```diff
@@ -922,6 +943,11 @@ def fit(self, df, **kwargs):
         -------
         The fitted Prophet object.
         """
+        if self.calculate_holidays:
+            # calculate years range
+            year_list = list(set([x.year for x in pd.to_datetime(df['ds'])]))
+            self.holidays = self.make_holidays(year_list=year_list, country_list=self.append_holidays)
+
```

Comment:
I think the best place for this is inside the method `make_holiday_features`. At the very start of that function we could do something like
```
all_holidays = self.holidays
if self.append_holidays is not None:
    year_list = list(set([x.year for x in dates]))
    append_holidays_df = self.make_holidays(year_list=year_list, country_list=self.append_holidays)
    all_holidays = pd.concat((all_holidays, append_holidays_df))
```
And then in the code that follows (L460 on this revision) , use `all_holidays` in the place of `self.holidays.

Like we talked about in the meeting, there is some extra work needed there to handle the fact that `append_holidays_df` there might return different sets of holidays depending on the years that are passed in. The `make_holiday_features` method is the place where this issue has to be addressed. Right now, the holidays dataframe is fixed as `self.holidays` and the only variable input is `dates`. These are used to produce a dictionary from holiday name to binary indicators (an element for each date), `expanded_holidays` in the code. `expanded_holidays` looks like
```
defaultdict(<function __main__.<lambda>>,
            {'playoff_delim_+0': array([0., 0., 0., ..., 0., 0., 0.]),
             'playoff_delim_+1': array([0., 0., 0., ..., 0., 0., 0.]),
             'superbowl_delim_+0': array([0., 0., 0., ..., 0., 0., 0.]),
             'superbowl_delim_+1': array([0., 0., 0., ..., 0., 0., 0.])})
```
We then do
```
holiday_features = pd.DataFrame(expanded_holidays)
```
Other code relies critically on the fact that the order of the columns in `holiday_features` will be the same every time we run it. So that is the thing we have to be sure of when adding these computed holidays: The columns and column order of `holiday_features` used in fitting the model must be the same as the columns/column order that we get from this in `predict`, when a different set of `dates` are passed in.

Reply:
I think we can ensure `holiday_features` keeps the same columns by tracking the names in `all_holidays['holiday']` when we first run `make_holiday_features`, and then in subsequent runs remove any rows from `all_holidays` whose holiday is not in the initial fitted list.

We can also sort `holiday_features` by column name prior to returning it, to be extra sure that things are staying correctly ordered. We just have to be sure to do the sort before `prior_scale_list` is computed.